### PR TITLE
Replaceable redeclare support

### PR DESCRIPTION
**Pull request by adrpo**

### CODE REVIEW

**Comment by casella** [comment]
@adrpo, can you quickly comment on the results of the PR testing? I can't say whether it is almost ok or if there is still some serious issue. 

Thanks!

**Comment by adrpo** [comment]
I ported a lot of stuff from master, but i forgot to update the testsuite as well. Should be quite easy to do. I'm working on it.

File: .gitignore [hunk 1]
```diff
@@ -21,5 +21,4 @@ OMPublicKey/
 OMPython/
 OMSetup/
 OpenModelicaSetup/
-<<<<<<< HEAD
-OMEncryption/
```
Comment (adeas31) [inline .gitignore:25]: Remove this git merge conflict marker.

**Comment by adeas31** [comment]
Can you please point out what are the main changes in OMEdit. So far all I can see is the rename from `Component` to `Element`.

File: .gitignore (hunk 1 above)
Comment (adrpo) [reply .gitignore:25]: That *was* removed, it was there (as you can see these are deletions, not additions).

Comment (adrpo) [reply .gitignore:25]: I will however add back OMEncryption to the ignore list.

Comment (adrpo) [reply .gitignore:25]: Nevermind, OMEncryption was already there, a bit further up.

**Comment by adrpo** [comment]
> @adrpo, I see there are 17 tests currently failing, but I can't make sense of all of them. Why should this PR have any effect at all on them, since it only involves API calls in interactive mode?
> 
> Is this because of the additional NF pushes?

Yes, that brought a lot more stuff with it. I'm mostly done with the fixes to the testsuite.

**Comment by casella** [comment]
OK, I'm looking forward to your next commit

**Comment by casella** [comment]
We're now down to 13 failing tests. Some I don't really understand, e.g.
Equation mismatch: diff says:
Equation mismatch: omc-diff says:
```
--------------------------------------Line 20: Lexical token differs:
expected: 
got:       
```
The only difference here is spaces (!)

**Comment by casella** [comment]
@adeas31, @mahge, @lochel, can you please help @adrpo fixing the remaining failing 13 tests?. 

We really, really, really need to get through this ASAP

**Comment by mahge** [comment]
@casella Sure. Just let us know which one Adrian is not working on. 

For what it is worth it is only 5 tests remaining from the libraries. They are just duplicated for gcc and clang giving you 10 total.

For the compliance tests. 2 out of 3 (`ArrayVectorVectorMul3 `and `OuterInPackage`) just need an update on the test configuration. They were set to fail because they did not have support. Now they are working so it is just a matter of updating the expected tests. 

The other `VectorizationMultiInputIllegal` needs a better error message and an actual failure in the front-end (instead of failing at codegen later). 

My guess is Adrian has probably already fixed these. If not I can take a look at `TestComplexSum2.mos`

**Comment by adrpo** [comment]
I have no idea why TestComplexSum2.mos fails. The backend doesn't seem to properly handle sum(...) but is rather strange as it works in 1.14 and 1.16 but not here. Would be good if you can have a look.
The one with spaces are because on commit we trim the trailing space but the diff algorithm expects it.

What I am fighting with it is to add more of the needed changes to this PR. I thought I cherry-picked them before but that seem to have failed and now I'm having issues with the merging. I'm missing two commits one with changes required for OMEdit and one with changes for omc. I am working to get those in asap.

**Comment by mahge** [comment]
@adrpo Ya TestComplexSum2 seemed odd to me as well because I remember removing '`omcQuote__text_`' completely from the compiler and replacing it by `Q_text_Q`. Maybe you missed the merge on that commit (maybe more commits).

**Comment by adrpo** [comment]
I already merged that commit from you, let me check if i need to do a push.

**Comment by adrpo** [comment]
The bad execution with -d=dumpindxdae vs the good one as diff is below.
Somehow it seems the inline or sum expansion does work properly and the sum remains.

```diff
50c50
< Variables (5)
---
> Variables (4)
52,56c52,55
< 1: c2.im:VARIABLE()  type: Real
< 2: c2.re:VARIABLE()  type: Real
< 3: c1[3].re:VARIABLE()  type: Real [3]
< 4: c1[2].re:VARIABLE()  type: Real [3]
< 5: c1[1].re:VARIABLE()  type: Real [3]
---
> 1: c2.re:VARIABLE()  type: Real
> 2: c1[3].re:VARIABLE()  type: Real [3]
> 3: c1[2].re:VARIABLE()  type: Real [3]
> 4: c1[1].re:VARIABLE()  type: Real [3]
59c58
< Equations (4, 5)
---
> Equations (4, 4)
64c63
< 4/4 (2): c2 = sum(c1[i] for i in 1:3)   [dynamic |0|0|0|0|]
---
> 4/4 (1): c2.re = c1[1].re + c1[2].re + c1[3].re   [dynamic |0|0|0|0|]
69c68
< 5 variables and equations
---
> 4 variables and equations
71,74c70,72
< var 2 is solved in eqn 4
< var 3 is solved in eqn 3
< var 4 is solved in eqn 2
< var 5 is solved in eqn 1
---
> var 2 is solved in eqn 3
> var 3 is solved in eqn 2
> var 4 is solved in eqn 1
79,82c77,80
< {3:3}
< {2:4}
< {1:5}
< ComplexEquation  {4:1, 2}
---
> {3:2}
> {2:3}
> {1:4}
> {4:1}
89c87
< Known variables only depending on parameters and constants - globalKnownVars (3)
---
> Known variables only depending on parameters and constants - globalKnownVars (4)
93a92
> 4: c2.im:VARIABLE(fixed = true )  = 0.0  type: Real
```

**Comment by mahge** [comment]
@adrpo Even if the backend did something wrong the string 'omcQuote' should not exist in the generated code. I can see in your commits that the changes from me (https://github.com/OpenModelica/OpenModelica/commit/065a64f74f20b2db2b26f53075ad76094332f6ee) are incorporated. But still `omcQuote` is there in the generated code for `TestComplexSum2.mos` . There is something not quite right there. 

Let's see if the last commits changed things. Otherwise there might be some odd merging issue there.

**Comment by adrpo** [comment]
This is what i get:

```
TestComplexSum2.c: In function 'TestComplexSum2_eqFunction_11':
TestComplexSum2.c:153:23: warning: implicit declaration of function 'omc_Complex_Q_2B_Q' [-Wimplicit-function-declaration]
         D_24tmpVar3 = omc_Complex_Q_2B_Q(threadData, omc_Complex(threadData, D_24tmpVar32Ere, D_24tmpVar32Eim), omc_Complex(threadData, D_24tmpVar22Ere, D_24tmpVar22Eim));
                       ^
TestComplexSum2.c:153:78: error: 'D_24tmpVar32Ere' undeclared (first use in this function)
         D_24tmpVar3 = omc_Complex_Q_2B_Q(threadData, omc_Complex(threadData, D_24tmpVar32Ere, D_24tmpVar32Eim), omc_Complex(threadData, D_24tmpVar22Ere, D_24tmpVar22Eim));
                                                                              ^
TestComplexSum2.c:153:78: note: each undeclared identifier is reported only once for each function it appears in
TestComplexSum2.c:153:95: error: 'D_24tmpVar32Eim' undeclared (first use in this function)
         D_24tmpVar3 = omc_Complex_Q_2B_Q(threadData, omc_Complex(threadData, D_24tmpVar32Ere, D_24tmpVar32Eim), omc_Complex(threadData, D_24tmpVar22Ere, D_24tmpVar22Eim));
                                                                                               ^
TestComplexSum2.c:153:137: error: 'D_24tmpVar22Ere' undeclared (first use in this function)
         D_24tmpVar3 = omc_Complex_Q_2B_Q(threadData, omc_Complex(threadData, D_24tmpVar32Ere, D_24tmpVar32Eim), omc_Complex(threadData, D_24tmpVar22Ere, D_24tmpVar22Eim));
                                                                                                                                         ^
TestComplexSum2.c:153:154: error: 'D_24tmpVar22Eim' undeclared (first use in this function)
         D_24tmpVar3 = omc_Complex_Q_2B_Q(threadData, omc_Complex(threadData, D_24tmpVar32Ere, D_24tmpVar32Eim), omc_Complex(threadData, D_24tmpVar22Ere, D_24tmpVar22Eim));
                                                                                                                                                          ^
<builtin>: recipe for target 'TestComplexSum2.o' failed
mingw32-make: *** [TestComplexSum2.o] Error 1
mingw32-make: *** Waiting for unfinished jobs....

```

**Comment by mahge** [comment]
@adrpo okay that makes sense 👌. I was looking at the jenkins output
https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-789/12/tests/

So maybe the last push added those changes. 

This I can investigate the backend issue.

**Comment by casella** [comment]
Thank you folks! I hope this is really the last hurdle...

**Comment by casella** [comment]
Still 15 tests failing...

**Comment by casella** [comment]
Back to three + 2 time 5.

Is there any way I could help with that?

**Comment by adrpo** [comment]
> Back to three + two times five.
> Is there any way I could help with that?
I'm not sure how for now. Myself I'm looking into why the merging of stuff from master causes these issues as it shouldn't!
Mahder is looking into why TestComplexSum2 is failing.

**Comment by casella** [comment]
Still 13 failures 😞

**Comment by mahge** [comment]
@caella The problems in this PR are caused by missing commits from master.
I am trying to rebase the changes from Adrian on master here https://github.com/OpenModelica/OpenModelica/pull/813. There is one small problem with compilation of OMEdit test. I will fix that. 

There are no more tests failing. The ones that you see on Jenkins are all trivial.

The only problem is that there might be some changes missing from Adrian. We will check that after.

**Comment by casella** [comment]
> @casella The problems in this PR are caused by missing commits from master.
> I am trying to rebase the changes from Adrian on master here #813. There is one small problem with compilation of OMEdit test. I will fix that.
> 
> There are no more tests failing. The ones that you see on Jenkins are all trivial.
> 
> The only problem is that there might be some changes missing from Adrian. We will check that after.

Good to hear that 😃 

However, this pull request was opened almost three weeks ago, and no matter how trivial these problems are, they are delaying the release and testing of 1.15.0-beta.1 indefinitely. We are only two months away from the target date to release 1.16.0. I hope the trivial problems can be resolved ASAP and that we can finally start letting end users test 1.15.0-beta.1. 

@adrpo, if the failing tests are trivial, can you force the merging of this PR, proceed with the release of 1.15.0-beta.1 and take care of them later for the final 1.15.0 release?

Thanks!

**Comment by mahge** [comment]
@casella The failing tests and the OMEdit testsuite issues are all fixed now. The CI is passing. However this does not mean that its all over. Adrian still needs to go through it and make sure nothing is missing from the original commit. And then incorporate any other changes he has made locally since he has probably added some more changes since the PR.

However, at least we now don't have to worry about missing commits from master which really complicated debugging.

**Comment by casella** [comment]
I hope this doesn't take another two months 😄

**Comment by casella** [comment]
> @casella The failing tests and the OMEdit testsuite issues are all fixed now. The CI is passing. 

How do you know that? The last run of the CI tests I see is 5 days old and failed 13 tests. Do I miss something? Should we give this PR another try at Jenkins, even without actually merging it?

> However this does not mean that its all over. Adrian still needs to go through it and make sure nothing is missing from the original commit. And then incorporate any other changes he has made locally since he has probably added some more changes since the PR.

I hope this doesn't take another two months 😄

**Comment by casella** [comment]
Is there any progress on this PR?

**Comment by adrpo** [comment]
Yes. I have been working on merging my missing changes and I'll push the commits in as soon as I can get it to compile.

**Comment by casella** [comment]
@adrpo is this going smoothly or are there any further issues?

**Comment by adrpo** [comment]
We'll open a new one for 1.16 (on master).

**Pull request closed without merging**